src/storage-client/src/sink.rs

```diff
@@ -430,105 +444,158 @@ async fn determine_latest_progress_record(
         // transactions, there might even be a lot of garbage at the end of the
         // topic or in between.
 
-        let mut tps = TopicPartitionList::new();
-        tps.add_partition(progress_topic, partition);
-        tps.set_partition_offset(progress_topic, partition, Offset::Beginning)?;
-
-        progress_client.assign(&tps).with_context(|| {
-            format!(
-                "Error seeking in progress topic {}:{}",
-                progress_topic, partition
-            )
-        })?;
-
-        let (lo, hi) = progress_client
-            .fetch_watermarks(progress_topic, 0, timeout)
+        // First, determine the current high water mark for the progress topic.
+        // This is the position our `progress_client` consumer *must* reach
+        // before we can conclude that we've seen the latest progress record for
+        // the specified `progress_key`. A safety argument:
+        //
+        //   * Our caller has initialized transactions before calling this
+        //     method, which prevents the prior incarnation of this sink from
+        //     committing any further progress records.
+        //
+        //   * We use `read_uncommitted` isolation to ensure that we fetch the
+        //     true high water mark for the topic, even if there are pending
+        //     transactions in the topic. If we used the `read_committed`
+        //     isolation level, we'd instead get the "last stable offset" (LSO),
+        //     which is the offset of the first message in an open transaction,
+        //     which might not include the last progress message committed for
+        //     this sink! (While the caller of this function has fenced out
+        //     older producers for this sink, *other* sinks writing using the
+        //     same progress topic might have long-running transactions that
+        //     hold back the LSO.)
+        //
+        //   * If another sink spins up and fences out the producer for this
+        //     incarnation of the sink, we may not see the latest progress
+        //     record... but since the producer has been fenced out, it will be
+        //     unable to act on our stale information.
```

Comment:
@sploiselle @bkirwi @petrosagg I did my best to tie together the existing safety argument with the improvements in this PR. Is this sufficiently clear?